vcard2to3.py

```diff
@@ -8,38 +8,63 @@
 class VCard:
     BEGIN = 'BEGIN:VCARD'
     END = 'END:VCARD'
-    FN = re.compile('FN[:;]')
+    N = re.compile('^N[:;]')
+    FN = re.compile('^FN[:;]')
     NICKNAME = re.compile('NICKNAME[:;]')
     def __init__(self, prune_empty=False):
         self.reset()
         self._prune_empty = prune_empty
     def reset(self):
         self.lines = []
         self._omit = False
+        self._n = None
         self._fn = None
-    def add(self, line):
-        self.lines.append(line)
+    def add(self, line, idx=None):
+        if idx is not None:
+            self.lines.insert(idx, line)
+        else:
+            self.lines.append(line)
+        if VCard.N.match(line):
+            self._n = line
         if VCard.FN.match(line):
             self._fn = line
     def omit(self):
         self._omit = True
     def valid(self):
-        return self._fn != None
-    def write(self, to):
-        if self._omit:
+        return self._n != None and self._fn != None
+    def repair(self):
+        if self.valid():
             return
         nick_idx = -1
-        for i,l in enumerate(self.lines[1:-1]):
+        n_idx = -1
+        fn_idx = -1
+        for i, l in enumerate(self.lines[1:-1]):
+            if VCard.N.match(l):
+                n_idx = i+1
+            if VCard.FN.match(l):
+                fn_idx = i+1
             if VCard.NICKNAME.match(l):
-                if self._fn and (self._fn[3:] == l[9:]):
-                    del(self.lines[i+1]) # NICKNAME == FN => remove nickname
-                else:
-                    nick_idx = i+1
-        if not self.valid() and nick_idx >= 0 and (not self._prune_empty or len(self.lines) > 4):
-            # no FN but NICKNAME found, use it
-            self.lines[nick_idx] = self.lines[nick_idx].replace('NICKNAME', 'FN', 1)
-            self._fn = self.lines[nick_idx]
-            nick_idx = -1
+                nick_idx = i+1
+        if not self.valid():
+            # insert the "equivalent" field after the existing one
+            # Arbitrary "convertion": whitespace in FN <-> ';' in N
+            # N and FN have precedence over NICKNAME
+            if n_idx >= 0 and self._fn is None:
+                self.add(self.lines[n_idx].replace('N', 'FN', 1).replace(';', ' ', 1), n_idx+1)
+            elif fn_idx >= 0 and self._n is None:
+                self.add(';'.join(self.lines[fn_idx].replace('FN', 'N', 1).split())+'\n', fn_idx+1)
+            elif nick_idx >= 0 and (not self._prune_empty or len(self.lines) > 4):
```

Comment:
Here there is something to arrange for handling the prune option correctly, and treat it coherently in the `N` case as well. I'm not sure about what the prune intent is.

Reply:
I removed some contacts with the prune option, I don't recall it exactly. In the current state, at least the help text is wrong, because it would omit contacts with only NICKNAME. I consider removing 'prune_empty' completely.